### PR TITLE
ElementSelection: New element selection context to support selecting elements

### PR DESCRIPTION
**Pull request by torkelo**
A quick proof of concept of an element selection solution. 

We are exploring an edit pane in dashboards that shows actions & options for any element or elements (multiple) selected. See first POC here: https://github.com/grafana/grafana/pull/96895 

In that POC, I used a very hacky global onClick handler and element attributes to mark elements as selectable and selected. However, because it relied on DOM element manipulation to add the selected state class, the selection state is visually lost whenever react re-renders. This PR is exploring a more robust react approach to element selection via a context. 

Context has 
* Current selected elements
* Select an element (or multiple via shift key) 
* Enable/disable selection (needed in dashboards when we go into edit mode or leave edit mode)

Point of feedback
* Is there a different / better way to do this? 
* How to clear selection, Currently I have a top level click handler and when an element is selected propagation is stopped. Another way would be to have a timer so that we do not clear selection if the selection was done in the last few milliseconds (somehow need to make sure the clear selection click is not triggered by an actual selection click). 
* Should the context come with a "Provider" implementation that handles the selection state & callback or should we leave that up to each implementation to handle? 
 
Outside the scope of this PR (So ignore)
* How to make it clear an element is selectable (cursor or other affordance)
* How to make an element that is both draggable and selectable and differentiate between the two (going to be a bit tricky with Panel but we solved it in the past) 

https://github.com/user-attachments/assets/663bae3b-19a1-4d27-80d4-d04c6aed5e37

### CODE REVIEW

File: public/app/features/sandbox/TestStuffPage.tsx [hunk 1]
```diff
@@ -14,26 +26,93 @@ export const TestStuffPage = () => {
...
+            <PanelChrome title={'Panel title 2'} selectionId="panel-2">
+              Hello
+            </PanelChrome>
+            <PanelChrome title={'Panel title 3'} selectionId="panel-3">
+              Hello
+            </PanelChrome>
+            <PanelChrome title={'Panel title 4'} selectionId="panel-4">
+              Hello
+            </PanelChrome>
+          </VizGridLayout>
```
Comment (torkelo) [inline public/app/features/sandbox/TestStuffPage.tsx:114]: Regarding this demo impl, thoughts
* Remove it 
* Move it to a storybook story for ElementSelectionContext 
* Keep it here

File: packages/grafana-ui/src/components/ElementSelectionContext/ElementSelectionContext.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    return {};
+  }
+
+  const context = useContext(ElementSelectionContext);
+  if (!context) {
+    return {};
+  }
+
+  const isSelected = context.selected.some((item) => item.id === id);
+  const onSelect = useCallback<React.PointerEventHandler>(
```
Comment (bfmatei) [inline packages/grafana-ui/src/components/ElementSelectionContext/ElementSelectionContext.tsx:37]: I think this is breaking the rules of hooks again.

**Review by bfmatei — APPROVED** [review]
I really like the overall implementation of this. Left two comments in the code but overall I really like it.

A couple of additional comments:
- let's avoid having the timer for selection clearing. I think we can improve the existing solution in the future if it becomes a problem
- I find providers pretty awful to work with, especially because they're causing more nesting in JSX which can lead to ugly code. However, one of my comments was about having a hook that consumes the context and encapsulates the entire logic of handling selection. Or we can have both and allow devs to use whatever feels better for them?

File: packages/grafana-ui/src/components/ElementSelectionContext/ElementSelectionContext.tsx (hunk 2 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/ElementSelectionContext/ElementSelectionContext.tsx:37]: not really, if is no context and a context is added there will be an unmount/remount 

so an early exit when there is no context is fine

File: public/app/features/sandbox/TestStuffPage.tsx [hunk 3]
```diff
@@ -14,26 +26,93 @@ export const TestStuffPage = () => {
   };
 
   const notifyApp = useAppNotification();
+  const [elementSelectionState, setElementSelectionState] = useState<ElementSelectionContextState>({
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/sandbox/TestStuffPage.tsx:29]: @bfmatei maybe, not sure what the real dashboard implementation will look like yet. But after we make that one we could we can see how re-usable it is

**Pull request merged**